### ISSUE TITLE
Make periodic NodeStatus updates cheaper

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#14733]
We're currently sending whole NodeStatuses only to update a Node heartbeat. This a big source of traffic in the cluster, and one of the possible causes of 1000-node cluster failures. We need to extract heartbeat to a new API 'Heartbeat' object with timestamp and object reference only, and make Kubelet/NodeController to use this object instead of NodeStatus to determine health of the Node.

cc: @wojtek-t @lavalamp @bgrant0607 @brendandburns @smarterclayton @timothysc @davidopp @fgrzadkowski

**Comment by wojtek-t** [comment]
cc @dchen1107

**Comment by smarterclayton** [comment]
Why not just add a heartbeat subresource to the node that performs the
operation?

**Comment by gmarek** [comment]
That's the plan for the implementation, but we still need to store it somewhere, hence the need for an API object.

**Comment by smarterclayton** [comment]
Why wouldn't the store continue to be on the node?  Virtual resources like scale / bind / etc already handle this.  Maybe that's what you're suggesting, just want to be sure.

**Comment by smarterclayton** [comment]
I.e. `PUT /nodes/foo/heartbeat` simply sets lastProbeTime and accepts the "Heartbeat" virtual sub resource you described.

**Comment by wojtek-t** [comment]
But this would require changing the implementation of how subresources are implemented. The problem is that currently we just send the whole object in that case (e.g. whole pod in case of binding). We would need efficient PATCH operation that would send only data that is changing (not the whole object).

**Comment by smarterclayton** [comment]
Are you looking to reduce UPDATE cost, or WATCH of updated node cost?

Binding is a minimal resource - it's just the object reference for the target node and the name of the pod.  For the former, you would POST/PUT heartbeat "lastProbeTime" to the apiserver, which would run a guaranteed update on node that only alters the lastProbeTime.  The kubelets would only send a minimal object, but anyone watching would still get the update.

For the latter, if the resources are truly split, wouldn't that require the node controller to fetch and watch two objects in order to stitch that data together?  I didn't think watch on node was the problem, which is why I assume you meant the former.

**Comment by wojtek-t** [comment]
Yes - I mostly meant the former.

For the watch - with the new watch in apiserver - we will read the data from etcd only once and then send it only to interested watchers (and there should be constant number of those).

So yes - I'm mostly worried about write path (not read path).

**Comment by smarterclayton** [comment]
Binding is modeled today as:

```
POST /v1/namespaces/<namespace>/pods/<pod>/binding {"kind":"Binding", "metadata": {"name": "<pod>"}, "to": {"name": "node1"}}
```

The result of that is persisted into the pod in etcd via the GuaranteedUpdate loop.

```
PUT /v1/namespaces/<namespace>/nodes/<node>/heartbeat {"kind":"Heartbeat", "metadata": {"name": "<node>"}, "lastProbeTime": "2015-10-09 12:03:00Z"}
```

Could be very similar.

**Comment by bgrant0607** [comment]
Have you tried PATCH?

**Comment by bgrant0607** [comment]
I assume that this is not for 1.1.

**Comment by gmarek** [comment]
Probably not, as we're already close to meeting our performance goals of v.1.1. It'll be (probably) necessary when we start to aim at more than 250 Nodes.

**Comment by bgrant0607** [comment]
If we were going to add a new subresource, it should match the conventions of other subresources (and API conventions generally) and we shouldn't call it Heartbeat -- it should be a condition-reporting API.

However, we shouldn't need to add a subresource just for efficiency reasons. We should make it efficient to patch and get subsets of resources. 

Is PATCH not implemented for node status? Strategic merge patch looks like it should work fine, though we might want to fix resourceVersion preconditions.

The ability to GET a subset of fields is covered by #1459.

**Comment by bgrant0607** [comment]
Also, do you have performance data showing how significant a problem this is?

**Comment by dchen1107** [comment]
I disagreed on introducing a Heartbeat object for this purpose. 

* From current issues reported, I don't think NodeStatus is the issue. Instead events and PodStatus should have more impact on this. 
* This is actually a generic issue we should solve. For example, we need to update lastProbeTime and lastTransitionTime for PodStatus. Extracting Heartbeat object doesn't solve this particular issue with the same root cause.

Can we use PATCH to do delta updates? Or introduce bulk updates?

**Comment by lavalamp** [comment]
-100 on heartbeat object.

Switch to PATCH. Even if we have to implement it.

**Comment by smarterclayton** [comment]
Oh yeah, patch is also way better.

On Sep 29, 2015, at 8:27 PM, Daniel Smith <notifications@github.com> wrote:

-100 on heartbeat object.

Switch to PATCH. Even if we have to implement it.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/14733#issuecomment-144146694>
.

**Comment by lavalamp** [comment]
I also want to add that while switching to PATCH would be nice, I don't believe at all that this is actually a problem. It's extra bandwidth to send the whole object, but we are not bandwidth limited, we are write QPS limited (right?) and that PATCH, the current PUT, and a hypothetical heartbeat object ALL cause the same number of etcd writes (and actually, the same exact data, since PATCH and heartbeat would still have to do a read/write cycle of the whole object).

So I expect changing this to make basically no difference to a cluster.

**Comment by dchen1107** [comment]
Bulk update can somewhat solve today's QPS limit issue (of course, if the limit is valid is a separate issue), but no matter using PATCH, PUT, heatbeat object and bulk update, we have the same amount etcd writes. :-)

**Comment by lavalamp** [comment]
Yes, exactly.

**Comment by quinton-hoole** [comment]
IMO this data should not reside on disk (i.e. etcd) at all.  It belongs in RAM, as it can be reconstructed at any time (within 5 seconds, per the current settings). Has anyone considered this design approach (not for 1.1, obviously).

**Comment by lavalamp** [comment]
Yes, we've discussed splitting spec and status into objects that can be
hosted in separate places. It's potentially a lot of work.

**Comment by timothysc** [comment]
Due to the CP write latency+encode+decode..., and the issues with in-memory state across load balanced api-server, has anyone started to discuss gossip protocols over heartbeats?  e.g. Riak+Consuls model.?.? 

Once we start hitting O(10^3+) even patched heartbeats will start to become painful.

**Comment by gmarek** [comment]
I got too preoccupied and forgot about this issue... The problem with PATCH is that, IIUC, it only helps if the network bandwidth is a problem, but I'm actually more worried about hammering etcd. With PATCH master will need to get an object (e.g. NodeStatus) form etcd, patch it and write it to the database.

I'm not sure if this won't give worse performance than current implementation (patch and resubmit logic in the API server instead of the Kubelet). If CPU will become a bottleneck for the API server we may be worse of than if we don't change anything.

What's nice about heartbeats is that they actually don't need to be patched - we can easily have an append only database for heartbeats only (which I guess should be much more efficient than the PATCH in etcd), and create an abstraction of 'one database' on API server level (i.e. make the separation of heartbeats an implementation detail).

**Comment by bgrant0607** [comment]
I'd much rather move all of status to a separate record, and then not persist it to disk. #8625

**Comment by wojtek-t** [comment]
@bgrant0607 - I think the problem with your proposal (although completely I agree that from the API & usability point of view it looks much better) is that NodeStatus would be almost the whole Node object (i.e. NodeSpec is pretty small object so there is no significant performance gain from splitting those two).

**Comment by smarterclayton** [comment]
Something to consider is whether we could have this be a concern of etcd -
a key prefix that went through raft consensus but did not require
persistence.  That avoids (in the short term) having to implement our own
consensus cache.

On Oct 7, 2015, at 12:14 PM, Brian Grant <notifications@github.com> wrote:

I'd much rather move all of status to a separate record, and then not
persist it to disk. #8625
<https://github.com/kubernetes/kubernetes/issues/8625>

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/14733#issuecomment-146248223>
.

**Comment by bgrant0607** [comment]
Has anyone tested etcd performance on tmpfs and possibly with relaxed consistency?

**Comment by quinton-hoole** [comment]
@bgrant0607 I'm busy trying to get such a cluster up as we speak.  Ran into some issues yesterday getting my large cluster up, but continuing today.

**Comment by davidopp** [comment]
@smarterclayton having non-persistent API objects was proposed in #6363 but it didn't go anywhere. (There was no concrete proposal as to how to implement it, and we wanted to wait to see if we needed it.) I like your suggestion a lot.

**Comment by gmarek** [comment]
I'm still not convinced that splitting heartbeat off to the separate object is a bad idea. Yes, it's less beautiful API, but as we noticed during Omega project, putting a lot of stuff in a huge objects that contain everything does not work very well for both: performance and development.

One reason why I think is bad is our watch semantics. I.e. we update Node objects quite often because of the heartbeat and we need to send updates to, e.g. all Kubelets, even though they don't care at all about it. 

I'd imagine that we can keep objects independent enough, so that updates get propagated only to the components that care. E.g. until recently we had this for Pods, which were updated only when something important changed (I'm not sure what's the current status of 'last probe time' effort).

**Comment by dchen1107** [comment]
We won't do anything with this for 1.2 release, and I don't think we plan to do anything for next release neither.

**Comment by davidopp** [comment]
I asked @wojtek-t yesterday if he thought this was needed for our scalability goals for the next release and he said probably not.

**Comment by davidopp** [comment]
Why did you close this? We might want to come back to it later.

**Comment by gmarek** [comment]
@bgrant0607  don't want to go this way. He has some other plan for separating Spec from Status, which would be enough.

**Comment by bgrant0607** [comment]
Changed the title of the issue to be more general.

**Comment by gmarek** [comment]
I believe there's a new effort around this started by @bgrant0607.
/close

**Comment by shyamjvs** [comment]
We probably should reopen this issue as we're seeing problems due to this in large clusters (~3k nodes) with heavy node objects (particularly due to images).
Etcd seems to be consuming too much memory in storing multiple versions of those objects.

**Comment by timothysc** [comment]
The image meta-data has always been the largest source of update.

**Comment by bgrant0607** [comment]
It seemed like we could have hashed the image names/digests into shorter strings, not unlike branch-predictor tables or even a bloom filter, which also would have addressed PII issues. It's just a performance hint.

**Comment by wojtek-t** [comment]
I thought a bit more about it and here are my thoughts:

1. I think there are two main very high-level approaches that we may potentially take here:
(a) Split the Node object in storage into two parts: "heartbeat" part and "the rest" part (obviously leaving everything as is at the api level)
(b) Somehow reuse the concept of "lease" from etcd v3 to implement cheap heartbeats at storage layer.

To elaborate a bit more on those two options:

(a) Split in storage:
- we need storage transactions to implement it (so this can be implemented only in etcd v3). However, since etcd v2 is deprecated (and we don't support 5k-node clusters with etcd v2 anyway), I think this is fine [if you are still using etcd v2, you will just have a single object in etcd]
- the typical crud operations seems relatively easy to implement here (it's just simple splitting/merging of two objects from storage into a single one)
- what is tricky is watch, because we obviously we don't want to change the existing semantic (so heartbeats are still watchable as they currently are)
That means, that whenever an update to heartbeat comes, we need to know the base Node object from exactly that time (and vice-versa). If we would be able to properly initialize the watch, it would work (keep the last node and heartbeast that was observed). The problem is how to properly initialize it (both part may be independently compacted in etcd transaction log).
The only thing that really comes to my mind is to disable ability to switch-off cacher for Nodes. That would mean that no matter what happens, watches for nodes are always served from cache. And with that, a strawman of what we could do is:
- when initializing cache: we List (in a single request) both "heartbeat" and "main" parts of nodes from etcd. Merging those gives us "current" state of the world.
- we start watching from there, and then we know on top of what an update should be applied (because we know previous version of full object).
In fact this isn't really different comparing to what we are doing now.


(b) change heartbeat semantic
**I'm not even convinced this is something we can do - it's not backward compatible from semantic perspective".
The high level idea is to make the heartbeat somehow leverage the concept of lease in etcd v3. However, since "refresh lease" concept is non-watchable in etcd, this would mean changing the semantic of heartbeats. Additionally, we won't be able to attach a least to "node" object in etcd, because expired lease would cause removing Node object, which is a complete change of semantic. So we would have need a separate object to which the lease would be attached. That exposes us to problems with watch already described in the first approach. Which means that it probably doesn't make sense to proceed with it, since on top of that, there are a bunch of many other problems here.

**Comment by bgrant0607** [comment]
A couple ideas:

First, reduce heartbeat write rate.

We must continue to write the lastHeartbeatTime field. This search gets thousands of hits:
https://github.com/search?utf8=%E2%9C%93&q=lastHeartbeatTime+kubernetes&type=Code

But we don't have to write it with the same frequency.

The Node controller is looking for signs that the Kubelet is still active. It could look for any recent Events posted by each Kubelet (using the reportingInstance info from the new Event API) and update the Node status appropriately, but much less frequently than the current heartbeat rate. The Kubelet could keep track of when it last sent any Event, and generate a dummy Heartbeat Event if it had nothing else to report. 

Node controller wouldn't depend on any particular event -- only the source would be considered. It wouldn't depend on the number of events -- just at least one would have to be observed within some time period. Some events could be lost and it would be ok. After the heartbeat was recorded, it would be ok for all of the events to be lost. I suggested events precisely because we already need to optimize storage and distribution of events. They could be stored in memory, and catastrophic loss of all the events would be ok -- we could conservatively introduce a longer observation period in the case that that occurred.

Both the Kubelet heartbeat frequency and node-controller eviction timeout are configurable. So we're effectively proposing a change of default values, introducing more health signals into node controller (which we've at least discussed since the beginning of the project), and shifting more of the responsibility of writing that field from kubelet to node controller (node controller already updates the condition when kubelet doesn't update it). We have to take into account that kubelets can be a few releases behind the node controller. I suggest making these changes gradually.

Second, reducing node status size.

@lavalamp and I have discussed reducing the size by performing lossy compression of the image data, such as smaller hashes, bloom filters, and such. It's just for scheduler placement optimization, doesn't have to be perfect, and just needs to be matched. Hashing the information also would mitigate PII issues to some degree.

**Comment by wojtek-t** [comment]
The current proposal is here: https://github.com/kubernetes/community/pull/2090/files

**Comment by shubheksha** [comment]
ping @wojtek-t / @wangzhen127 - is this on track for 1.13?

**Comment by wangzhen127** [comment]
Yes

**Comment by AishSundar** [comment]
@wangzhen127 can you provide a list of pending PRs for this issue and an ETA for when we can expect all the related PRs merged? Can you please give an apporpriate kind/ label for this issue (eg: feature, bug, cleanup etc)

We are nearing code slush for 1.13 next Friday 11/9 and Code freeze in 2 weeks. If you think this issue isn't critical for 1.13 and work might extend beyond code slush please work with Niko Pen to move this out. thanks

**Comment by wangzhen127** [comment]
/kind feature
This feature is critical for 1.13. I think it is in good progress.

Code
- https://github.com/kubernetes/kubernetes/pull/69753 (expect to merge this week)
- https://github.com/kubernetes/kubernetes/pull/70034 (expect to merge next week)

Doc
- https://github.com/kubernetes/website/pull/10699 (expect to merge next week)

Test Infra
- https://github.com/kubernetes/test-infra/pull/9942 (expect to merge next week)

**Comment by AishSundar** [comment]
Thanks so much @wangzhen127 for the detailed PR list. helps us to track much better

@tfogo for Doc related PR

**Comment by wangzhen127** [comment]
Implementation is finished.
Documentation is merged into https://github.com/kubernetes/website/blob/dev-1.13/content/en/docs/concepts/architecture/nodes.md
Tests are being tracked by the following CI jobs:
sig-node
- https://k8s-testgrid.appspot.com/sig-node-kubelet#node-kubelet-alpha
- https://k8s-testgrid.appspot.com/sig-node-containerd#e2e-gci-alpha-features

google-gce
- https://k8s-testgrid.appspot.com/google-gce#gci-gce-alpha-features

google-gke
- https://k8s-testgrid.appspot.com/google-gke#gci-gke-alpha-features
- https://k8s-testgrid.appspot.com/google-gke#cos-containerd-gke-1.13-alpha-cluster-features
- https://k8s-testgrid.appspot.com/google-gke#cos-gke-1.13-alpha-cluster-features

/close

**Comment by wojtek-t** [comment]
Let's keep that open until this is enabled by default (currently, this is an Alpha feature, thus disabled by default).

We will do a bit more testing and we planning to enable it by default for 1.14.

**Comment by liggitt** [comment]
can this be closed, now that #72096 is merged?

**Comment by wojtek-t** [comment]
It's a matter of promoting the API and the feature to GA.
But I guess we don't need this bug to be opened and we should open another one for it...
So I think it's fine

**Comment by wojtek-t** [comment]
On a related note - I just started working on promoting API to v1 - will send out a PR later today or tomorrow.

**Comment by nikopen** [comment]
@wojtek-t as #72239 has been merged, can this be closed?

**Comment by wojtek-t** [comment]
The feature is not in GA (it will be in Beta in 1.14), but I think we can close it - there is a corresponding issue in enhancements repo.
